### PR TITLE
[FONT][WIN32SS] Improve font coordinate transformation

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Improve coordinate transformation except `LOGFONT.lfOrientation` in font/text rendering.
JIRA issue: [CORE-15554](https://jira.reactos.org/browse/CORE-15554), [CORE-15623](https://jira.reactos.org/browse/CORE-15623)

### CODE REVIEW

**Comment by katahiromz** [comment]
Underline position is OK.
underline-ok

**Comment by katahiromz** [comment]
CORE-11848 testcase is OK.
testcase-ok

**Comment by katahiromz** [comment]
Why shifted to lower right?
why-shifted

**Comment by katahiromz** [comment]
Proof of Concept: https://jira.reactos.org/secure/attachment/50739/TypeTest4.7z
Test Program: https://jira.reactos.org/secure/attachment/50740/TypeTest4NoFreeType.zip
Win10 Screenshots: https://jira.reactos.org/secure/attachment/50743/TypeTest4-win10.zip
AFTER Screenshots: https://jira.reactos.org/secure/attachment/50744/TypeTest4-ros-after.zip

**Comment by katahiromz** [comment]
AFTER:

**Comment by katahiromz** [comment]
Win10:

ROS AFTER:

In ReactOS, setting matrix (0, 1, 1, 0) to the world fails.

**Comment by katahiromz** [comment]
Win10:

ROS AFTER:

Oh no! The world transformation and logical scaling have to be separated. It looks like mixture.

**Comment by katahiromz** [comment]
I did what I had to do. Please review me.

**Comment by katahiromz** [comment]
This PR don't fix CORE-14994, CORE-15478, and CORE-15303.

**Comment by yagoulas** [comment]
Test results are here : https://www.reactos.org/testman/compare.php?ids=65128,65131
Tomorrow I'll review the changes

File: win32ss/gdi/ntgdi/freetype.c [hunk 1]
```diff
@@ -5612,9 +5612,7 @@ GreExtTextOutW(
     {
         Start.x = pdcattr->ptlCurrent.x;
         Start.y = pdcattr->ptlCurrent.y;
-    }
-    else
-    {
+    } else {
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c]: OK.

**Comment by yagoulas** [comment]
@tkreuzer what do you think about the changes in XFORMOBJ_iSetXform? I don't really know the details about this module so I can't really review it.

**Comment by katahiromz** [comment]
`FLOATOBJ` is a special type of `float` in the kernel. To manipulate these types, we have to use `FLOATOBJ_*` functions. `M11 * M22 - M12 * M21` is the determinant of the XFORM. If the determinant was zero, then the matrix is not invertible.

https://math.stackexchange.com/questions/355644/what-does-it-mean-to-have-a-determinant-equal-to-zero

If the transformation matrix was not invertible, then the image has zero volume. As observation, `SetWorldTransform` function refuse the non-invertible matrix.

**Comment by katahiromz** [comment]
For example,
```txt
1, 1,
1, 1
```
is a non-invertible matrix.

**Comment by katahiromz** [comment]
It seems nobody else can review this PR. I will force commit after one week.

**Comment by JoachimHenze** [comment]
katahiromz, we create a big diff here to official freetype and our existing rostests seem to be more or less immune to those changes. Can you add a few automated tests to rostests maybe, that can be run against windows as well and that demonstrate the improvements?
This would serve two things:
1.) strengthen the the trust in your changes
2.) In case someone later aims to sync something with official freetype, those tests can prevent losing, what you achieved here.
Freetype lib in kernel is very sensitive code from security pov.
We should aim to have more automated tests for changes to kernel-font-rendering-code.

My personal manual test results with patch-state 2019-01-28 applied to 0.4.12-dev-466-g08d72bd:
-I can confirm your PR fixes regression CORE-15623 but we shouldn't forget that this bug is also a recently introduced regression caused by changes to that code, that did not happen with unchanged freetype. (An automated test triggering that difference would be nice as well)
-PR has no effect on CORE-12378
-PR has no effect on CORE-15319

Thank you!

**Pull request merged**